File: ArchitectureHW2/lowerTriangularMatrix.cpp
#include "lowerTriangularMatrix.h"
#include <cstdio>
#include <random>

// Base constructor.
LowerTriangularMatrix::LowerTriangularMatrix() {
    size_ = 0;
    elems_count = 0;
}

// Determined size constructor.
LowerTriangularMatrix::LowerTriangularMatrix(int size) {
    size_ = size;
    elems_count = (size * (size + 1)) / 2;
}

// File input.
void LowerTriangularMatrix::Read(FILE *input) {
    for (int i = 0; i < elems_count; ++i) {
        fscanf(input, "%lf", &elems[i]);
    }
}

// Random input.
void LowerTriangularMatrix::RandomRead() {
    size_ = 0;
    size_ = rand() % 100 + 1;
    elems_count = (size_ * (size_ + 1)) / 2;
    for (int i = 0; i < elems_count; ++i) {
        elems[i] = (double) (rand() / (double) RAND_MAX / 1000000) * rand();
    }
}

// Output.
void LowerTriangularMatrix::Out(FILE *output) {
    fprintf(output, "%s: %d\n", "Size", size_);
    fprintf(output, "%s: %f\n", "Average", getAverage());
    fprintf(output, "%s:\n", "Lower-triangular matrix");

    int counter = 0;
    for (int i = 0; i < size_; ++i) {
        for (int j = 0; j < size_; ++j) {
            // Only lower-triangular elements != 0.
            if (i < j) {
                fprintf(output, "%s ", "0");
            } else {
                fprintf(output, "%f ", elems[counter]);
                ++counter;
            }
        }
        fprintf(output, "\n");
    }
}

// Getting average of all elements.
double LowerTriangularMatrix::getAverage() {
    // Total sum.
    double sum = 0;

    for (int i = 0; i < elems_count; ++i) {
        sum = sum + elems[i];
    }

    return sum / size_ / size_;
}

File: Rust/src/matrices/lower_triangular_matrix.rs
use crate::matrices::base_matrix::BaseMatrix;
use rand::rngs::ThreadRng;
use rand::Rng;
use std::fs::File;
use std::io::Write;

pub struct LowerTriangularMatrix {
    pub size: usize,
    pub elems_count: usize,
    pub elems: Vec<f64>,
}

impl BaseMatrix for LowerTriangularMatrix {
    fn input(&mut self, info: &mut [String], index: &mut usize) {
        self.size = info[*index].parse::<usize>().unwrap();
        *index += 1;
        self.elems_count = self.size * (self.size + 1) / 2;
        let matr_info: Vec<String> = info[*index]
            .split(' ')
            .map(|s: &str| s.to_string())
            .collect();
        *index += 1;

        for i in 0..self.elems_count {
            self.elems.push(matr_info[i].parse::<f64>().unwrap());
        }
    }

    fn random_input(&mut self) {
        let mut gen: ThreadRng = rand::thread_rng();
        let _rand_num: usize = gen.gen();

        self.size = gen.gen_range(1..101);
        self.elems_count = self.size * (self.size + 1) / 2;
        for _i in 0..self.elems_count {
            self.elems.push(gen.gen_range(-1000.0..1000.0));
        }
    }

    fn output(&self, file: &mut File) -> Result<(), std::io::Error> {
        file.write_all(
            format!(
                "Lower-triangular matrix with size of {} and average element {}:\n",
                self.size,
                self.get_average()
            )
            .as_bytes(),
        )?;

        for i in 0..self.size {
            for j in 0..self.size {
                if i < j {
                    file.write_all("0 ".as_bytes())?;
                    continue;
                }
                file.write_all(format!("{}, ", self.elems[i]).as_bytes())?;
            }
            file.write_all(b"\n")?;
        }

        Ok(())
    }

    fn get_average(&self) -> f64 {
        let mut sum: f64 = 0_f64;
        for i in 0..self.elems_count {
            sum += self.elems[i];
        }
        return sum / self.size as f64 / self.size as f64;
    }
}
